http4s/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorApp.scala

```diff
@@ -12,47 +14,70 @@ import org.typelevel.log4cats.Logger
 import org.typelevel.log4cats.slf4j.Slf4jLogger
 
 import java.net.InetSocketAddress
-import scala.concurrent.duration.DurationLong
+import scala.concurrent.duration.{DurationLong, FiniteDuration}
 
 object CollectorApp {
 
-  implicit private def unsafeLogger: Logger[IO] =
-    Slf4jLogger.getLogger[IO]
+  implicit private def unsafeLogger[F[_]: Sync]: Logger[F] =
+    Slf4jLogger.getLogger[F]
 
-  def run(): IO[ExitCode] =
-    buildHttpServer().use(_ => IO.never).as(ExitCode.Success)
+  def run[F[_]: Async](mkGood: Resource[F, Sink[F]], mkBad: Resource[F, Sink[F]]): F[ExitCode] = {
+    val resources = for {
+      bad <- mkBad
+      good <- mkGood
+      _ <- withGracefulShutdown(610.seconds) {
+        buildHttpServer[F](new CollectorRoutes[F](good, bad).value)
+      }
+    } yield ()
 
-  private def buildHttpServer(): Resource[IO, Server] =
+    resources.surround(Async[F].never[ExitCode])
+  }
+
+  private def withGracefulShutdown[F[_]: Async, A](delay: FiniteDuration)(resource: Resource[F, A]): Resource[F, A] =
+    for {
+      a <- resource
+      _ <- Resource.onFinalizeCase {
+        case Resource.ExitCase.Canceled =>
+          Logger[F].warn(s"Shutdown interrupted. Will continue to serve requests for $delay") >>
+            Async[F].sleep(delay)
```

Comment:
I was thinking the finalizer of the Resource would flush the sink and call shutdown.  For example, the kinesis implementation of this sink would look like:

```scala
val acquire = Sync[F].delay {
  KinesisSink.createAndInitialize(....)
}
Resource.make(acquire)(sink => Sync[F].delay(sink.shutdown()))
```

...but that's a sink-specific decision about how to shut down cleanly.

I did also consider leaving `shutdown()` as part of the trait, and then calling it from `CollectorApp`.  But... I just didn't like that so much.  What do you think?

> Is it also a moment to set health endpoint to return false?

In the previous collector, we implemented that behaviour just in case we needed it in Kubernetes setups.  But so far we have not needed.  We found that Kubernetes pod rotation is working very nicely even without flipping the health endpoint to false.  We have disabled it everywhere.

So I would vote to keep things simple and not change the health endpoint to false.

Reply:
You're right, that's pretty much whole point of a `Resource` 🤦. I agree with everything you've written!